File: rd-net/RdFramework/Impl/RdSet.cs
using System;
using System.Collections;
using System.Collections.Generic;
using JetBrains.Annotations;
using JetBrains.Collections.Synchronized;
using JetBrains.Collections.Viewable;
using JetBrains.Diagnostics;
using JetBrains.Lifetimes;
using JetBrains.Rd.Base;
using JetBrains.Rd.Util;
using JetBrains.Serialization;
using JetBrains.Threading;

namespace JetBrains.Rd.Impl
{
  public class RdSet<T> : RdReactiveBase, IViewableSet<T> where T: notnull
  {
    private readonly IViewableSet<T> mySet;

    public RdSet(CtxReadDelegate<T> readValue, CtxWriteDelegate<T> writeValue)
    {
      ValueCanBeNull = false;
      ReadValueDelegate = readValue;
      WriteValueDelegate = writeValue;

      mySet = new ViewableSet<T>(new SynchronizedSet<T>() /*to have thread-safe print*/);
    }


    #region Serializers

    public CtxReadDelegate<T> ReadValueDelegate { get; private set; }
    public CtxWriteDelegate<T> WriteValueDelegate { get; private set; }

    public static RdSet<T> Read(SerializationCtx ctx, UnsafeReader reader)
    {
      return Read(ctx, reader, Polymorphic<T>.Read, Polymorphic<T>.Write);
    }

    public static RdSet<T> Read(SerializationCtx ctx, UnsafeReader reader, CtxReadDelegate<T> readValue, CtxWriteDelegate<T> writeValue)
    {
      var id = reader.ReadRdId();
      return new RdSet<T>(readValue, writeValue).WithId(id);
    }

    public static void Write(SerializationCtx ctx, UnsafeWriter writer, RdSet<T> value)
    {
      Assertion.Assert(!value.RdId.IsNil);
      writer.Write(value.RdId);
    }
    #endregion


    #region Mastering

    public bool IsMaster = false;

    #endregion


    #region Init
    
    public bool OptimizeNested { [PublicAPI] get; set; }

    protected override void PreInit(Lifetime lifetime, IProtocol parentProto)
    {
      base.PreInit(lifetime, parentProto);
      if (lifetime.IsNotAlive)
        return;

      parentProto.Wire.Advise(lifetime, this);
    }

    protected override void Init(Lifetime lifetime, IProtocol proto, SerializationCtx ctx)
    {
      base.Init(lifetime, proto, ctx);
      if (lifetime.IsNotAlive)
        return;
      
      using (UsingLocalChange())
      {
        Advise(lifetime, it =>
        {
          if (lifetime.IsNotAlive)
            return;
          
          if (!IsLocalChange) return;

          proto.Wire.Send(RdId, (stream) =>
          {
            stream.Write((int)it.Kind);
            WriteValueDelegate(ctx, stream, it.Value);


            SendTrace?.Log($"{this} :: {it.Kind} :: {it.Value.PrintToString()}");
          });
        });
      }
    }

    public override RdWireableContinuation OnWireReceived(Lifetime lifetime, IProtocol proto, SerializationCtx ctx, UnsafeReader stream, UnsynchronizedConcurrentAccessDetector? detector)
    {
      var kind = (AddRemove) stream.ReadInt();
      var value = ReadValueDelegate(ctx, stream);
      ReceiveTrace?.Log($"{this} :: {kind} :: {value.PrintToString()}");

      return new RdWireableContinuation(lifetime, detector, () =>
      {
        using (UsingDebugInfo())
        {
          switch (kind)
          {
            case AddRemove.Add:
              mySet.Add(value);
              break;

            case AddRemove.Remove:
              mySet.Remove(value);
              break;

            default:
              throw new ArgumentOutOfRangeException();
          }
        }
      });
    }

    #endregion


    #region Read delegation

    IEnumerator IEnumerable.GetEnumerator()
    {
      return mySet.GetEnumerator();
    }

    public IEnumerator<T> GetEnumerator()
    {
      return mySet.GetEnumerator();
    }

    public bool Contains(T item)
    {
      return mySet.Contains(item);
    }

    public void CopyTo(T[] array, int arrayIndex)
    {
      mySet.CopyTo(array, arrayIndex);
    }


    public int Count => mySet.Count;

    public bool IsReadOnly => mySet.IsReadOnly;


    public ISource<SetEvent<T>> Change => mySet.Change;

    #endregion



    #region Write delegation

    public bool Remove(T item)
    {
      using (UsingLocalChange())
        return mySet.Remove(item);
    }
    
    // ReSharper disable once AssignNullToNotNullAttribute
    #if NET35
    public
    #endif
    void
      #if !NET35
      ICollection<T>.
      #endif
      Add(T item)
    {
      AssertNullability(item);
      using (UsingLocalChange())
        mySet.Add(item);
    }

    #if !NET35
    public bool Add(T item)
    {
      AssertNullability(item);
      using (UsingLocalChange())
        return mySet.Add(item);
    }

    public void ExceptWith(IEnumerable<T> other)
    {
      using (UsingLocalChange())
        mySet.ExceptWith(other);
    }

    public void IntersectWith(IEnumerable<T> other)
    {
      using (UsingLocalChange())
        mySet.IntersectWith(other);
    }
    
    public void SymmetricExceptWith(IEnumerable<T> other)
    {
      using (UsingLocalChange())
        mySet.SymmetricExceptWith(other);
    }

    public void UnionWith(IEnumerable<T> other)
    {
      using (UsingLocalChange())        
        mySet.UnionWith(other);
    }
    #endif

    public void Clear()
    {
      using (UsingLocalChange()) 
        mySet.Clear();
    }
    
    #endregion



    #region ISet Read delegation

    #if !NET35
    public bool IsProperSubsetOf(IEnumerable<T> other) => mySet.IsProperSubsetOf(other);

    public bool IsProperSupersetOf(IEnumerable<T> other) => mySet.IsProperSupersetOf(other);

    public bool IsSubsetOf(IEnumerable<T> other) => mySet.IsSubsetOf(other);

    public bool IsSupersetOf(IEnumerable<T> other) => mySet.IsSupersetOf(other);

    public bool Overlaps(IEnumerable<T> other) => mySet.Overlaps(other);

    public bool SetEquals(IEnumerable<T> other) => mySet.SetEquals(other);
    #endif

    #endregion

    

    public void Advise(Lifetime lifetime, Action<SetEvent<T>> handler)
    {
      using (CreateAssertThreadingCookie(null))
      using (UsingDebugInfo())
        mySet.Advise(lifetime, handler);
    }


    protected override string ShortName => "set";
    public override void Print(PrettyPrinter printer)
    {
      base.Print(printer);
      if (!printer.PrintContent) return;

      printer.Print(" [");
      if (Count > 0) printer.Println();

      using (printer.IndentCookie())
      {
        foreach (var v in this)
        {
          v.PrintEx(printer);
          printer.Println();
        }
      }
      printer.Println("]");
    }
  }

}

File: rd-kt/rd-framework/src/main/kotlin/com/jetbrains/rd/framework/impl/RdSet.kt
package com.jetbrains.rd.framework.impl

import com.jetbrains.rd.framework.*
import com.jetbrains.rd.framework.base.*
import com.jetbrains.rd.util.collections.SynchronizedSet
import com.jetbrains.rd.util.lifetime.Lifetime
import com.jetbrains.rd.util.reactive.*
import com.jetbrains.rd.util.string.PrettyPrinter
import com.jetbrains.rd.util.string.print
import com.jetbrains.rd.util.string.printToString
import com.jetbrains.rd.util.trace

@Suppress("UNUSED_PARAMETER")
open class RdSet<T : Any> constructor(val valueSerializer: ISerializer<T>, private val set: ViewableSet<T>)
: RdReactiveBase(), IMutableViewableSet<T> {

    companion object : ISerializer<RdSet<*>> {
        fun<T: Any> read(ctx: SerializationCtx, stream: AbstractBuffer, valueSerializer: ISerializer<T>): RdSet<T> = RdSet(valueSerializer).withId(RdId.read(stream))

        override fun read(ctx: SerializationCtx, buffer: AbstractBuffer): RdSet<*> {
            return read(ctx, buffer, Polymorphic())
        }

        override fun write(ctx: SerializationCtx, buffer: AbstractBuffer, value: RdSet<*>) {
            value.rdid.write(buffer)
        }
    }

    override fun deepClone(): IRdBindable = RdSet(valueSerializer).also { for (elem in set) { it.add(elem.deepClonePolymorphic()) } }

    var optimizeNested : Boolean = false

    override fun preInit(lifetime: Lifetime, proto: IProtocol) {
        super.preInit(lifetime, proto)

        proto.wire.advise(lifetime, this)
    }

    override fun init(lifetime: Lifetime, proto: IProtocol, ctx: SerializationCtx) {
        super.init(lifetime, proto, ctx)

        localChange { advise(lifetime) lambda@{ kind, v ->
            if (!isLocalChange) return@lambda

            proto.wire.send(rdid) { buffer ->
                buffer.writeEnum(kind)
                valueSerializer.write(ctx, buffer, v)

                logSend.trace { "set `$location` ($rdid) :: $kind :: ${v.printToString()} "}
            }
        }}
    }

    override fun onWireReceived(proto: IProtocol, buffer: AbstractBuffer, ctx: SerializationCtx, dispatchHelper: IRdWireableDispatchHelper) {
        val kind = buffer.readEnum<AddRemove>()
        val v = valueSerializer.read(ctx, buffer)

        logReceived.trace { "onWireReceived:: $this :: $kind :: ${v.printToString()}" }
        dispatchHelper.dispatch {
            logReceived.trace { "dispatched:: $this :: $kind :: ${v.printToString()}" }
            when (kind) {
                AddRemove.Add -> set.add(v)
                AddRemove.Remove -> set.remove(v)
            }
        }
    }

    constructor(valueSerializer: ISerializer<T> = Polymorphic<T>()) : this(valueSerializer, ViewableSet(SynchronizedSet()))

    override val size: Int
        get() = set.size
    override val change: ISource<IViewableSet.Event<T>>
        get() = set.change

    override fun contains(element: T): Boolean = set.contains(element)

    override fun containsAll(elements: Collection<T>): Boolean = set.containsAll(elements)

    override fun isEmpty(): Boolean = set.isEmpty()

    override fun print(printer: PrettyPrinter) {
        super.print(printer)
        printer.print(" [")
        if (!isEmpty()) printer.println()

        printer.indent {
            forEach {
                it.print(printer)
                printer.println()
            }
        }
        printer.print("]")
    }



    override fun add(element: T): Boolean {
        return localChange { set.add(element) }
    }

    override fun addAll(elements: Collection<T>): Boolean {
        return localChange { set.addAll(elements) }
    }

    override fun iterator(): MutableIterator<T> {
        val delegate = set.iterator()
        return object : MutableIterator<T> by delegate {
            override fun remove() {
                localChange { delegate.remove() }
            }
        }
    }

    override fun remove(element: T): Boolean {
        return localChange { set.remove(element) }
    }

    override fun removeAll(elements: Collection<T>): Boolean {
        return localChange { set.removeAll(elements) }
    }

    override fun retainAll(elements: Collection<T>): Boolean {
        return localChange { set.retainAll(elements) }
    }

    override fun clear() = localChange { set.clear() }

    override fun advise(lifetime: Lifetime, handler: (IViewableSet.Event<T>) -> Unit) {
        if (isBound) assertThreading()
        set.advise(lifetime, handler)
    }


}
